File: microsynth/microsynth/report/label_manager/label_manager.py
# Copyright (c) 2024, Microsynth
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
import re
import json
from microsynth.microsynth.seqblatt import set_status


def get_columns(filters):
    return [
        {"label": _("Label Barcode"), "fieldname": "label_id", "fieldtype": "Data", "width": 100 },
        {"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 75 },
        {"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 75 },
        {"label": _("Customer Name"), "fieldname": "customer_name", "fieldtype": "Data", "width": 175 },
        {"label": _("Contact"), "fieldname": "contact", "fieldtype": "Link", "options": "Contact", "width": 70 },
        {"label": _("Sales Order"), "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order", "width": 125 },
        {"label": _("Web Order ID"), "fieldname": "web_order_id", "fieldtype": "Data", "width": 95 },
        {"label": _("Item Code"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 75 },
        {"label": _("Registered"), "fieldname": "registered", "fieldtype": "Check", "width": 80 },
        {"label": _("Registered To"), "fieldname": "registered_to", "fieldtype": "Link", "options": "Contact", "width": 100 },
        {"label": _("Sequencing Label"), "fieldname": "name", "fieldtype": "Link", "options": "Sequencing Label", "width": 115 }
    ]


def get_data(filters):
    if not filters:
        return []
    conditions = ""
    
    if filters.get('contact'):
        conditions += f"AND `tabSequencing Label`.`contact` = '{filters.get('contact')}'"
    if filters.get('registered_to'):
        conditions += f"AND `tabSequencing Label`.`registered_to` = '{filters.get('registered_to')}'"
    if filters.get('customer'):
        conditions += f"AND `tabSequencing Label`.`customer` = '{filters.get('customer')}'"
    if filters.get('customer_name'):
        conditions += f"AND `tabSequencing Label`.`customer_name` LIKE '{filters.get('customer_name')}'"
    if filters.get('sales_order'):
        conditions += f"AND `tabSequencing Label`.`sales_order` = '{filters.get('sales_order')}'"
    if filters.get('web_order_id'):
        conditions += f"AND `tabSales Order`.`web_order_id` = '{filters.get('web_order_id')}'"
    if filters.get('item_code'):
        conditions += f"AND `tabSequencing Label`.`item` = '{filters.get('item_code')}'"
    if filters.get('registered'):
        conditions += f"AND `tabSequencing Label`.`registered` = 1"
    if filters.get('from_barcode') and filters.get('to_barcode'):
        if filters.get('from_barcode').isnumeric() and filters.get('to_barcode').isnumeric():
            from_barcode = filters.get('from_barcode')
            to_barcode = filters.get('to_barcode')
            if len(from_barcode) != len(to_barcode):
                frappe.throw("From Barcode and To Barcode need to have the same length. Please use leading zeros if necessary.")
            barcode_list = ','.join(f'"{i:0{len(to_barcode)}d}"' for i in range(int(from_barcode), int(to_barcode) + 1))
            conditions += f"AND `tabSequencing Label`.`label_id` IN ({barcode_list})"
            #conditions += f"AND `tabSequencing Label`.`label_id` BETWEEN '{filters.get('from_barcode')}' AND '{filters.get('to_barcode')}'"  # leads to false positive search results
        else:
            from_prefix = ''.join([i for i in filters.get('from_barcode') if not i.isdigit()])
            to_prefix = ''.join([i for i in filters.get('to_barcode') if not i.isdigit()])
            if from_prefix != to_prefix:
                frappe.throw("From Barcode and To Barcode need to have the same Prefix.")
            from_barcode = re.sub("[^0-9]", "", filters.get('from_barcode'))
            to_barcode = re.sub("[^0-9]", "", filters.get('to_barcode'))
            if len(from_barcode) != len(to_barcode):
                frappe.throw("From Barcode and To Barcode need to have the same length.")
            barcode_list = ','.join(f'"{to_prefix}{i:0{len(to_barcode)}d}"' for i in range(int(from_barcode), int(to_barcode) + 1))
            conditions += f"AND `tabSequencing Label`.`label_id` IN ({barcode_list})"
    elif (filters.get('from_barcode') or filters.get('to_barcode')) and not conditions:
    #    frappe.throw( _("For using from and to barcode, please set both filters.") )
        return []
    elif filters.get('from_barcode') or filters.get('to_barcode'):
    #    frappe.msgprint( _("From and to barcode are both required, this filter is being ignored.") )
        return []
    
    sql_query = f"""
        SELECT
            `tabSequencing Label`.`name`,
            `tabSequencing Label`.`status`,
            `tabSequencing Label`.`registered`,
            `tabSequencing Label`.`label_id`,
            `tabSequencing Label`.`item` AS `item_code`,
            `tabSequencing Label`.`customer`,
            `tabSequencing Label`.`customer_name`,
            `tabSequencing Label`.`sales_order`,
            `tabSales Order`.`web_order_id`,
            `tabSequencing Label`.`contact`,
            `tabSequencing Label`.`registered_to`
        FROM `tabSequencing Label`
        LEFT JOIN `tabSales Order` ON `tabSales Order`.`name` = `tabSequencing Label`.`sales_order`
        WHERE TRUE
            {conditions}
        ORDER BY `tabSequencing Label`.`label_id`;
        """
    data = frappe.db.sql(sql_query, as_dict=True)
    return data


def execute(filters):
    columns, data = get_columns(filters), get_data(filters)
    return columns, data


def create_label_log(from_status, to_status, reason, description, content_str, filters):
    label_log_doc = frappe.get_doc({
        'doctype': 'Label Log',
        'user': frappe.session.user,
        'from_status': from_status,
        'to_status': to_status,
        'reason': reason,
        'description': description,
        'labels': content_str
    })
    label_log_doc.insert()
    label_log_doc.update(filters)
    # add Web Order ID from Sales Order if missing
    if filters.get('sales_order') and not filters.get('web_order_id'):
        label_log_doc.web_order_id = frappe.get_value("Sales Order", filters.get('sales_order'), 'web_order_id')
    # add Customer from Sales Order if missing
    if filters.get('sales_order') and not filters.get('customer'):
        label_log_doc.customer = frappe.get_value("Sales Order", filters.get('sales_order'), 'customer')
    # add Customer Name from Sales Order if missing
    if filters.get('sales_order') and not filters.get('customer_name'):
        label_log_doc.customer_name = frappe.get_value("Sales Order", filters.get('sales_order'), 'customer_name')
    # add Item Code if missing
    if not filters.get('item_code') and content_str:
        content = json.loads(content_str)
        labels = content.get('labels')
        item_codes = set()
        for l in labels:
            item_codes.add(l.get('item_code'))
        if len(item_codes) == 1:
            [item_code] = item_codes  # tuple unpacking verifies the assumption that the set contains exactly one element (raising ValueError if it has too many or too few elements)
            label_log_doc.item_code = item_code
    label_log_doc.save()
    frappe.db.commit()


@frappe.whitelist()
def lock_labels(content_str, filters, reason, description):
    """
    Set label status to 'locked'. Labels must be a list of dictionaries 
    (see `set_status` function).
    If locking was successfull, create a Label Log.
    """
    if type(content_str) == str:
        content = json.loads(content_str)
    else:
        content = content_str
    if type(filters) == str:
        filters = json.loads(filters)
    response = set_status("locked", content.get("labels"))
    if response['success']:
        create_label_log('unused', 'locked', reason, description, content_str, filters)
    return response


@frappe.whitelist()
def set_labels_unused(content_str, filters, reason, description):
    """
    Set label status to 'unused'. Labels must be a list of dictionaries 
    (see `set_status` function).
    If setting to unused was successfull, create a Label Log.
    """
    if type(content_str) == str:
        content = json.loads(content_str)
    else:
        content = content_str
    if type(filters) == str:
        filters = json.loads(filters)
    response = set_status("unused", content.get("labels"))
    if response['success']:
        create_label_log('locked', 'unused', reason, description, content_str, filters)
    return response

File: microsynth/microsynth/report/label_manager/label_manager.js
// Copyright (c) 2024, Microsynth
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Label Manager"] = {
    "filters": [
        {
            "fieldname": "contact",
            "label": __("Contact"),
            "fieldtype": "Link",
            "options": "Contact"
        },
        {
            "fieldname": "registered_to",
            "label": __("Registered To"),
            "fieldtype": "Link",
            "options": "Contact"
        },
        {
            "fieldname": "customer",
            "label": __("Customer ID"),
            "fieldtype": "Link",
            "options": "Customer"
        },
        {
            "fieldname": "customer_name",
            "label": __("Customer Name"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "from_barcode",
            "label": __("From Barcode"),
            "fieldtype": "Data",
            "on_change": function() {
                if ((frappe.query_report.get_filter_value("from_barcode")) && (!frappe.query_report.get_filter_value("to_barcode"))) {
                    frappe.query_report.set_filter_value("to_barcode", frappe.query_report.get_filter_value("from_barcode"));
                }
                frappe.query_report.refresh();
            }
        },
        {
            "fieldname": "to_barcode",
            "label": __("To Barcode"),
            "fieldtype": "Data",
            "on_change": function() {
                if ((frappe.query_report.get_filter_value("to_barcode")) && (!frappe.query_report.get_filter_value("from_barcode"))) {
                    frappe.query_report.set_filter_value("from_barcode", frappe.query_report.get_filter_value("to_barcode"));
                }
                frappe.query_report.refresh();
            }
        },
        {
            "fieldname": "label_status",
            "label": __("Label Status"),
            "fieldtype": "Select",
            "options": "unknown\nunused\nsubmitted\nreceived\nprocessed\nlocked"
        },
        {
            "fieldname": "sales_order",
            "label": __("Sales Order"),
            "fieldtype": "Link",
            "options": "Sales Order"
        },
        {
            "fieldname": "web_order_id",
            "label": __("Web Order ID"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "item_code",
            "label": __("Item Code"),
            "fieldtype": "Link",
            "options": "Item",
            get_query: function () {
                return {
                    'filters': {
                        'is_sales_item': 1
                        // TODO: only show Items that have a Label Range associated
                    }
                };
            }
        },
        {
            "fieldname": "registered",
            "label": __("Registered"),
            "fieldtype": "Check"
        }
    ],
    "onload": (report) => {
        hide_chart_buttons();
        hide_column_filters();
        report.page.add_inner_button( __("Lock Labels"), function() {
            if (frappe.query_report.data.length == 0) {
                frappe.msgprint("No Labels to lock.");
            } else if (frappe.query_report.data.length > 1000) {
                frappe.msgprint("Unable to lock more than 1000 Labels at once.");
            } else {
                var labels_to_lock = [];
                // check that all labels are unused
                var all_labels_unused = true;
                for (var i = 0; i < frappe.query_report.data.length; i++) {
                    if (frappe.query_report.data[i].status != "unused") {
                        frappe.msgprint("The Sequencing Label " + frappe.query_report.data[i].name + " with Barcode " + frappe.query_report.data[i].label_id + " has Status " + frappe.query_report.data[i].status + ". Unable to set status to 'locked'. No Label status was changed. Please contact IT App if you have a valid use case.");
                        all_labels_unused = false;
                        break;
                    }
                    labels_to_lock.push({
                        'label_id': frappe.query_report.data[i].label_id,
                        'item_code': frappe.query_report.data[i].item_code
                    });
                }
                if (all_labels_unused) {
                    // ask for a reason
                    frappe.prompt(
                        [
                            {
                                'fieldname': 'user',
                                'fieldtype': 'Link',
                                'label': __('User'),
                                'options': 'User',
                                'default': frappe.session.user,
                                'read_only': 1
                            },
                            {
                                'fieldname': 'reason',
                                'fieldtype': 'Select',
                                'label': __('Reason'),
                                'options': 'Payment of additional reactions\nLost',
                                'description': 'Please contact IT App if additional reasons are required.',
                                'reqd': 1
                            },
                            {
                                'fieldname': 'description',
                                'fieldtype': 'Text',
                                'label': __('Description')
                            }
                        ],
                        function(values) {
                            var my_filters = {};
                            for (var i = 0; i < frappe.query_report.filters.length; i++) {
                                my_filters[frappe.query_report.filters[i].fieldname] = frappe.query_report.filters[i].value;
                            }
                            // trigger label locking
                            frappe.call({
                                'method': "microsynth.microsynth.report.label_manager.label_manager.lock_labels",
                                'args': {
                                    'content_str': {'labels': labels_to_lock},
                                    'filters': my_filters,
                                    'reason': values.reason,
                                    'description': values.description || ''
                                },
                                'freeze': true,
                                'freeze_message': __('Locking ' + frappe.query_report.data.length + ' Labels ...'),
                                'callback': function(r) {
                                    if (r.message.success) {
                                        frappe.show_alert('Locked Labels');
                                        frappe.click_button('Refresh');
                                    } else {
                                        frappe.throw('Unable to lock Labels:<br>' + r.message.message + '<br><br>No Labels were locked.');
                                    }
                                }
                            });
                        },
                        __('Lock ' + frappe.query_report.data.length + ' Labels?'),
                        __('Lock')
                    )
                }
            }
        }).addClass("btn-primary");

        report.page.add_inner_button( __("Set Labels unused"), function() {
            if (frappe.query_report.data.length == 0) {
                frappe.msgprint("No Labels to set unused.");
            } else if (frappe.query_report.data.length > 100) {
                frappe.msgprint("Unable to set more than 100 Labels at once to unused.");
            } else {
                var labels_to_set_unused = [];
                // check that all labels are locked
                var all_labels_locked = true;
                for (var i = 0; i < frappe.query_report.data.length; i++) {
                    if (frappe.query_report.data[i].status != "locked") {
                        frappe.msgprint("The Sequencing Label " + frappe.query_report.data[i].name + " with Barcode " + frappe.query_report.data[i].label_id + " has Status " + frappe.query_report.data[i].status + ". Unable to set status to 'unused'. No Label status was changed. Please contact IT App if you have a valid use case.");
                        all_labels_locked = false;
                        break;
                    }
                    labels_to_set_unused.push({
                        'label_id': frappe.query_report.data[i].label_id,
                        'item_code': frappe.query_report.data[i].item_code
                    });
                }
                if (all_labels_locked) {
                    // ask for a reason
                    frappe.prompt(
                        [
                            {
                                'fieldname': 'user',
                                'fieldtype': 'Link',
                                'label': __('User'),
                                'options': 'User',
                                'default': frappe.session.user,
                                'read_only': 1
                            },
                            {
                                'fieldname': 'reason',
                                'fieldtype': 'Select',
                                'label': __('Reason'),
                                'options': 'Arrived at Customer after considered lost',
                                'description': 'Please contact IT App if additional reasons are required.',
                                'reqd': 1
                            },
                            {
                                'fieldname': 'description',
                                'fieldtype': 'Text',
                                'label': __('Description')
                            }
                        ],
                        function(values) {
                            var my_filters = {};
                            for (var i = 0; i < frappe.query_report.filters.length; i++) {
                                my_filters[frappe.query_report.filters[i].fieldname] = frappe.query_report.filters[i].value;
                            }
                            // set Labels to status unused
                            frappe.call({
                                'method': "microsynth.microsynth.report.label_manager.label_manager.set_labels_unused",
                                'args':{
                                    'content_str': {'labels': labels_to_set_unused},
                                    'filters': my_filters,
                                    'reason': values.reason,
                                    'description': values.description || ''
                                },
                                'freeze': true,
                                'freeze_message': __('Setting ' + frappe.query_report.data.length + ' Labels to unused ...'),
                                'callback': function(r) {
                                    if (r.message.success) {
                                        frappe.show_alert('Set Labels to unused');
                                        frappe.click_button('Refresh');
                                    } else {
                                        frappe.throw('Unable to set Labels to unused:<br>' + r.message.message + '<br><br>No Labels were set to unused.');
                                    }

                                }
                            });
                        },
                        __('Set ' + frappe.query_report.data.length + ' Labels to unused?'),
                        __('Set unused')
                    )
                }
            }
        }).addClass("btn-primary");
    }
};

function hide_column_filters() {
    let container = document.getElementsByClassName("page-content");
    const hide_column_filter_style = document.createElement("style");
    hide_column_filter_style.innerHTML = `
        .dt-header .dt-row[data-is-filter] {
          display: none !important;
        }
    `
    for (let i = 0; i < container.length; i++) {
        container[i].appendChild(hide_column_filter_style);
    }
}
